## Python Basics

### Question 1 (5 points)
**Complete the function `str_cmp` to return `True` if two strings are the same irrespective of lowercase/uppercase and whitespaces. Else, return `False`.**

In [ ]:
def str_cmp(str_1, str_2):
    # YOUR CODE HERE
     str_1_new = str_1.lower().replace(" ", "")
     str_2_new = str_2.lower().replace(" ", "")
     return str_1_new==str_2_new




In [ ]:
# TEST CASES -- DO NOT MODIFY

# Test Case 1
assert str_cmp("csce", "csce") == True, "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert str_cmp("csce", "csce412") == False, "Failed test case 2"
print("Passed test case 2\n" + "="*20)

# Test Case 3
assert str_cmp("MaChine LeArnIng", "machine learning") == True, "Failed test case 3"
print("Passed test case 3\n" + "="*20)

# Test Case 4
assert str_cmp("Machine Learning!", "Machine Learning") == False, "Failed test case 4"
print("Passed test case 4\n" + "="*20)

# Test Case 5
assert str_cmp("  Machine Learning", "machine    Learning ") == True, "Failed test case 5"
print("Passed test case 5\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
Passed test case 3
Passed test case 4
Passed test case 5
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 2 (5 points)
**Complete the function `power_list` in which each element is raised by its index to create a new list. Return the sum of the elements of the new list.**  

Example:  
given list: [1, 2, 3, 4]  
new list: [1, 2, 9, 64]  
sum of the elements of new list: 76

In [ ]:
def power_list(arr):
    ## YOUR CODE HERE
    sum=0
    for i in range(len(arr)):
        sum+=arr[i]**i
    return sum




In [ ]:
# TEST CASES -- DO NOT MODIFY

# Test Case 1
assert power_list([i for i in range(5)]) == 289, "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert power_list([-i for i in range(10, -1, -1)]) == 48, "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 3 (10 points)
**Complete the function `create_stats` that returns the mean and median values of the values arranged in a nested list.**

In [ ]:

def create_stats(universities):
    """
    Return a tuple of (mean students enrolled, median students enrolled, mean tuition, median tuition)
    Return a tuple of (mean cars sold, median cars sold, mean MSRP, median MSRP)
    """
    # YOUR CODE HERE
    import statistics

    stuorcar= [x[1] for x in universities]
    tuiormsrp = [x[2] for x in universities]
    mean_stuorcar = statistics.mean(stuorcar)
    median_stuorcar = statistics.median(stuorcar)
    mean_tuiormsrp = statistics.mean(tuiormsrp)
    median_tuiormsrp = statistics.median(tuiormsrp)
    return (mean_stuorcar, median_stuorcar, mean_tuiormsrp, median_tuiormsrp)





In [ ]:
# TEST CASES -- DO NOT MODIFY

# [[university name, students enrolled, tuition]]
universities = [
    ["Harvard", 19627, 39849],
    ["Massachusetts Institute of Technology", 10566, 40732],
    ["Princeton", 7806, 37000],
    ["Rice", 5870, 35551],
    ["Stanford", 19539, 40569],
    ["Yale", 11706, 40500],
]

# [[model, cars sold, MSRP]]
cars = [
    ["Ford F-series", 52314, 35000],
    ["Toyota RAV4", 41145, 30000],
    ["Honda CR-V", 31530, 32000],
    ["Toyota Camry", 32959, 28000],
    ["Honda Civic", 28318, 23000]
]

# Test Case 1
assert create_stats(universities) == (12519.0, 11136.0, 39033.5, 40174.5), "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert create_stats(cars) == (37253.2, 32959, 29600.0, 30000), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


# NumPy Basics

In [77]:
import numpy as np

### Question 4 (5 points)
**Complete the function `matrix_reshape` to reshape a matrix to the target dimensions given by `n_rows` and `n_cols`.**

In [ ]:
def matrix_reshape(mat, n_rows, n_cols):
    # YOUR CODE HERE
    A=mat.reshape((n_rows,n_cols))
    return A



In [ ]:
# TEST CASES -- DO NOT MODIFY

A = np.array([
    [3, -4, 2, 9],
    [6, 1, -5, 2],
    [-7, 8, -1, 6]
])

# Test Case 1
assert (matrix_reshape(np.arange(15), 3, 5) == np.array([[ 0,  1,  2,  3,  4], [ 5,  6,  7,  8,  9],
                                                            [10, 11, 12, 13, 14]])).all(), "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert (matrix_reshape(A, 2, 6) == np.array([[ 3, -4,  2,  9,  6,  1],
       [-5,  2, -7,  8, -1,  6]])).all(), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 5 (5 points)
**Complete the function `dot_product` that returns the dot product of two matrices. Dot product is also known as matrix multiplication.**

In [ ]:
def dot_product(mat_1, mat_2):
    # YOUR CODE HERE
    A= np.dot(mat_1,mat_2)
    return A





In [ ]:
# TEST CASES -- DO NOT MODIFY

A = np.array([
    [3, -4, 2, 9],
    [6, 1, -5, 2],
    [-7, 8, -1, 6]
])

# Test Case 1
assert dot_product(A[0,:], A[1,:]) == 22, "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert (dot_product(A[0,:], A.T) == np.array([110,  22,  -1])).all(), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

# Test Case 3
assert (dot_product(A, A.T) == np.array([[110,  22,  -1], [ 22,  66, -17],
                                               [ -1, -17, 150]])).all(), "Failed test case 3"
print("Passed test case 3\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
Passed test case 3
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 6 (5 points)
**Complete the function `cross_product` that returns the cross product of two matrices.**

In [ ]:
def cross_product(mat_1, mat_2):
    # YOUR CODE HERE
    return np.cross(mat_1,mat_2)




In [ ]:
# TEST CASES -- DO NOT MODIFY

A = np.array([
    [3, -4, 2, 9],
    [6, 1, -5, 2],
    [-7, 8, -1, 6]
])

# Test Case 1
assert cross_product(A[0,:2], A[:2,1]) == -13, "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert cross_product(A[:2,:2], A[2:,2:])[0] == 14, "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 7 (5 points)
**Complete the function `inverse` that computes the inverse of a matrix.**

In [ ]:
def inverse(mat):
    # YOUR CODE HERE
    I = np.linalg.inv(mat)
    return I



In [ ]:
# TEST CASES -- DO NOT MODIFY

A = np.array([
    [6, 1, 1],
    [4, -2, 5],
    [2, 8, 7]
])

# Test Case 1
assert np.allclose(inverse(A), np.array([[ 0.17647059, -0.00326797, -0.02287582],
       [ 0.05882353, -0.13071895,  0.08496732],
       [-0.11764706,  0.1503268 ,  0.05228758]])), "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert (inverse(A[:2,:2]) == np.array([[ 0.125 ,  0.0625],
       [ 0.25  , -0.375 ]])).all(), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 8 (10 points)
**Complete the function `subtract_mean` that subtracts the row-wise mean from each row of a matrix.**

In [ ]:
def subtract_mean(mat):
    ## YOUR CODE HERE
    x = np.mean(mat, axis=1, keepdims=True)
    ans = mat - x
    return ans



In [ ]:
A = np.array([
    [3, -4, 2, 9],
    [6, 1, -5, 2],
    [-7, 8, -1, 6]
])

# Test Case 1
assert (subtract_mean(np.arange(12).reshape(4,3)) == np.array([[-1.,  0.,  1.], [-1.,  0.,  1.],
       [-1.,  0.,  1.], [-1.,  0.,  1.]])).all(), "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert (subtract_mean(A) == np.array([[0.5, -6.5, -0.5,  6.5], [5. ,  0. , -6. ,  1. ],
        [-8.5,  6.5, -2.5,  4.5]])).all(), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


### Question 9 (10 points)
**Complete the function `moving_average` that computes averages over a sliding window of size `window_size` for a 1-D array.**

In [96]:
def moving_average(arr, window_size):
    # YOUR CODE HERE
    #I used stackoverflow : "is there faster way to convolve" article
    ans = np.convolve(arr, np.ones([window_size]), mode='valid')
    return ans/window_size




In [95]:
# Test Case 1
assert (moving_average(np.arange(10), 3) == np.arange(8)+1).all(), "Failed test case 1"
print("Passed test case 1\n" + "="*20)

# Test Case 2
assert (moving_average(np.array([-10, 5, 6, -4, 7, 9]), 2) ==
        np.array([-2.5, 5.5, 1, 1.5, 8])).all(), "Failed test case 2"
print("Passed test case 2\n" + "="*20)

print("*"*29 + "\n**  PASSED ALL TEST CASES  **\n" + "*"*29)

Passed test case 1
Passed test case 2
*****************************
**  PASSED ALL TEST CASES  **
*****************************


# Pandas Basics

In [ ]:
import pandas as pd


**Data**   


In [ ]:
sdata = {'From': ['delhi', 'Lucknow', 'delhi',
                   'Mumbai', 'Chennai'],
        'To': ['Hyderabad', 'Kochi', 'Chandigarh', 'Hyderabad', 'Kolkata'],
        'FlightNumber': [10045, np.nan, 10065, np.nan, 10085],
        'Delay_1': [23, np.nan, 24, 13, 67],
        'Delay_2': [47, np.nan, 43, np.nan, 32],
        'Delay_3': [np.nan, np.nan, 87, np.nan, np.nan],
        'Airline': ['Air India', 'Indigo', 'Go Air',
                    'Indigo', 'Vistara']}

In [ ]:
df = pd.DataFrame(sdata)

In [ ]:
df.head()

From          To  FlightNumber  Delay_1  Delay_2  Delay_3    Airline
0    delhi   Hyderabad       10045.0     23.0     47.0     42.0  Air India
1  Lucknow       Kochi       10055.0     42.0     42.0     42.0     Indigo
2    delhi  Chandigarh       10065.0     24.0     43.0     87.0     Go Air
3   Mumbai   Hyderabad       10075.0     13.0     42.0     42.0     Indigo
4  Chennai     Kolkata       10085.0     67.0     32.0     42.0    Vistara

### Question 10 (5 points)
**Calculate the number of unique cities in `df`.**

In [ ]:
## YOUR CODE HERE
x = pd.concat([df['From'], df['To']]).nunique()
print(x)





8


### Question 11 (5 points)
**Compute the no. of flights flying out of `Delhi`.**

In [ ]:
## YOUR CODE HERE
noflights= df['From'].value_counts()['delhi']
print(noflights)




2


### Question 12 (5 points)
**Which cities are served by `Indigo`?**

In [ ]:
## YOUR CODE HERE
byindigo = df[df['Airline'] == 'Indigo'][['From', 'To']].stack().unique()
print(byindigo)




['Lucknow' 'Kochi' 'Mumbai' 'Hyderabad']


### Question 13 (5 points)
**Calculate the total number of missing values (they are `NaN`) in `df`.**

In [ ]:
## YOUR CODE HERE
y=df.isna().sum().sum()
print(y)




9


### Task 14 (10 points)
**Compute the average delay, disregarding any `NaN` values, across all flights. Subsequently, update `df` by substituting all `NaN` values within the `Delay_1`, `Delay_2`, and `Delay_3` columns with this computed average delay.**

In [ ]:
# YOUR CODE HERE
ad=df[['Delay_1', 'Delay_2', 'Delay_3']].stack().mean()
df['Delay_1'].fillna(ad, inplace=True)
df['Delay_2'].fillna(ad, inplace=True)
df['Delay_3'].fillna(ad, inplace=True)




In [ ]:
df.head()

From          To  FlightNumber  Delay_1  Delay_2  Delay_3    Airline
0    delhi   Hyderabad       10045.0     23.0     47.0     42.0  Air India
1  Lucknow       Kochi           NaN     42.0     42.0     42.0     Indigo
2    delhi  Chandigarh       10065.0     24.0     43.0     87.0     Go Air
3   Mumbai   Hyderabad           NaN     13.0     42.0     42.0     Indigo
4  Chennai     Kolkata       10085.0     67.0     32.0     42.0    Vistara

### Task 15 (10 points)
**Fill in missing values in the `FlightNumber` column. The `FlightNumber` is expected to increase by 10 with each subsequent row. For instance, at row indices 1 and 3, the `FlightNumber` should be 10055 and 10075 respectively. Adjust the `df` to incorporate these changes and convert the column to integer type.**

In [ ]:
flight1st = df['FlightNumber'].dropna().iloc[0]
flightnums = np.arange(flight1st, flight1st + 10 * len(df), 10)
df['FlightNumber'] = flightnums


In [ ]:
df.head()

From          To  FlightNumber  Delay_1  Delay_2  Delay_3    Airline
0    delhi   Hyderabad       10045.0     23.0     47.0     42.0  Air India
1  Lucknow       Kochi       10055.0     42.0     42.0     42.0     Indigo
2    delhi  Chandigarh       10065.0     24.0     43.0     87.0     Go Air
3   Mumbai   Hyderabad       10075.0     13.0     42.0     42.0     Indigo
4  Chennai     Kolkata       10085.0     67.0     32.0     42.0    Vistara

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   From          5 non-null      object 
 1   To            5 non-null      object 
 2   FlightNumber  5 non-null      float64
 3   Delay_1       5 non-null      float64
 4   Delay_2       5 non-null      float64
 5   Delay_3       5 non-null      float64
 6   Airline       5 non-null      object 
dtypes: float64(4), object(3)
memory usage: 408.0+ bytes
